daemon/policy.go

```diff
@@ -302,6 +308,86 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (newRev uint
 	return newRev, nil
 }
 
+type policyAddQueueOptions struct {
+	rules   policyAPI.Rules
+	opts    *AddOptions
+	context *policyQueueContext
+	reFetch bool
+}
+
+type policyQueueContext struct {
+	policyRevision uint64
+	err            error
+	done           chan bool
+	totalTime      spanstat.SpanStat
+	queueTime      spanstat.SpanStat
+}
+
+// policyAddToQueue adds the request to policyAdd in the queue and return a
+// context that will be updated when the done channel is closed.
+func (d *Daemon) policyAddToQueue(rules policyAPI.Rules, opts *AddOptions, reFetch bool) *policyQueueContext {
+	result := &policyQueueContext{
+		policyRevision: 0,
+		err:            nil,
+		done:           make(chan bool),
+	}
+	result.totalTime.Start()
+	result.queueTime.Start()
+
+	d.policyAddQueueRequests <- policyAddQueueOptions{
+		rules:   rules,
+		opts:    opts,
+		reFetch: reFetch,
+		context: result,
+	}
+
+	return result
+}
+
+// PolicyQueueRequestInitilize initialize the daemon request channel and start
+// in backgrond the queue worker
+func (d *Daemon) PolicyQueueRequestInitilize(ctx context.Context) {
+	if d.policyAddQueueRequests == nil {
+		d.policyAddQueueRequests = make(chan policyAddQueueOptions, POLICY_REQUEST_CHANNEL_SIZE)
+	}
+	go d.policyQueueWorker(ctx)
+}
+
+// policyQueueWorker is a queue worker that taking care of PolicyAdd for
+// DNSProxy and Kubernetes watcher to avoid multiples races and lock all policy
+// repo.
+// In case that the channel event has a refetch set to true, the rule will be
+// retrieved from policyRepo to make sure that the newest version is installed
+// correctly and old rules do not overwritte the new ones.
+func (d *Daemon) policyQueueWorker(ctx context.Context) {
+	log.Info("Initialize Policy Queue Worker")
+	for {
+		select {
+		case policyRequest, _ := <-d.policyAddQueueRequests:
+			policyRequest.context.queueTime.End(true)
+			var rules []*policyAPI.Rule
+			rules = policyRequest.rules
+			if policyRequest.reFetch {
+				newRules := policyAPI.Rules{}
+				d.policy.Mutex.Lock()
+				for _, sourceRule := range policyRequest.rules {
+					repoRules := d.policy.SearchRLocked(sourceRule.Labels)
```

Comment:
Hi @tgraf 

To do that, as you mentioned, we need to hold the policyRepo lock in the PolicyAdd just before the StartManageDNS, so rules are retrieved and pushed correctly. 

https://github.com/cilium/cilium/blob/1010c58f4c2da2862c6159c73b331f28c613f14b/daemon/policy.go#L197-L230

The reason to do the calculation of MarkFQDNRules is to retrieve the IPS from the fqdn/cache.go, after that we need to create the CIDR identities, where a allocator lock is in place, meanwhile we hold the PolicyRepo.   

I mentioned this last Tuesday, but you mentioned the channel way, if we move the lock the issue I think that it'll be almost fixed, need to see all the corner cases, but I think that issue will be fixed.

Reply:
@tgraf ^^^

Reply:
The reason refetching is critical is that the uuid->rule mapping is a secondary store of rules. If an update happens to the policy repo, but a DNS response calls UpdateGenerateDNS before the new rules have been entered into the RuleGen.Regexpmap then the wrong rule copy will be passed along to PolicyAdd*.  This means we need to refetch at the last moment, while holding the policy repo lock.
Since we’re moving StartManageDNSName to within a repository.Lock block, maybe we can also hold that lock when calling UpdateGenerateDNS in NotifyOnDNSMessage. We would also call PolicyAddToQueue while holding it. This ensures that rules enter the policyAdd queue in order, and that a k8s or API update happens after a DNS update, or the DNS update fetches the correct version of the rules. This way, if a k8s or API update changes rules with ToFQDNs rules we will not end up in an inconsistent situation.

Reply:
@eloycoto @raybejjani Discussing during the Monday meeting. We decided that it is fine to merge with the re-fetch delayed into the policy addition and then revisit once @ianvernon's work around policy addition serialization has been merged.